pkg/controllers/provisioning/provisioner.go

```diff
@@ -89,15 +90,27 @@ func (p *Provisioner) provision(ctx context.Context) (err error) {
 	logging.FromContext(ctx).Infof("Batched %d pods in %s", len(items), window)
 	// Filter pods
 	pods := []*v1.Pod{}
+
+	// Multiple reconciliations for the same pod can occur while we are batching
```

Comment:
> Multiple reconciliations for the same pod can occur

Single reconciliation is an invariant of the codebase. It looks like this might be violated from your observations. Better to chase it down than plug the gap here, I think.

Reply:
I'm going to try to reproduce it on a real EKS cluster instead of just in my simulator sometime when I get chance.  Hopefully it's just an artifact of a not real-world system I was testing with.

Reply:
It appears to be a real issue and is 100% reproducible for me with the script attached, but still may not occur too often in real life since it requires the pods to be modified during the batching window.  This is with 0.6.1 on a real EKS cluster.

The script below scales a deployment to 10 pods, but in the controller logs the same pods were batched multiple times leading to binding errors and extra nodes being created:

```log

2022-02-11T02:46:05.398Z	INFO	controller.provisioning	Batched 25 pods in 4.257326352s	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:05.513Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 8 pod(s) with instance type option(s) [m5zn.3xlarge m5n.4xlarge m6a.4xlarge m5.4xlarge m5d.4xlarge m5dn.4xlarge m5ad.4xlarge m4.4xlarge m5a.4xlarge m6i.4xlarge i3.4xlarge]	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.775Z	INFO	controller.provisioning	Launched instance: i-0ab47b054785d5bcd, hostname: ip-192-168-47-85.ec2.internal, type: m6a.4xlarge, zone: us-east-1b, capacityType: spot	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.788Z	ERROR	controller.provisioning	Failed to bind default/inflate-85675546d-jqxq8 to ip-192-168-47-85.ec2.internal, pods "inflate-85675546d-jqxq8" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.789Z	ERROR	controller.provisioning	Failed to bind default/inflate-7bfb54665-4lsk9 to ip-192-168-47-85.ec2.internal, pods "inflate-7bfb54665-4lsk9" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.790Z	ERROR	controller.provisioning	Failed to bind default/inflate-7bfb54665-4skzb to ip-192-168-47-85.ec2.internal, pods "inflate-7bfb54665-4skzb" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.790Z	ERROR	controller.provisioning	Failed to bind default/inflate-7bfb54665-9mpt2 to ip-192-168-47-85.ec2.internal, pods "inflate-7bfb54665-9mpt2" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.790Z	ERROR	controller.provisioning	Failed to bind default/inflate-7bfb54665-2f9ws to ip-192-168-47-85.ec2.internal, pods "inflate-7bfb54665-2f9ws" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.790Z	ERROR	controller.provisioning	Failed to bind default/inflate-7bfb54665-wzwh2 to ip-192-168-47-85.ec2.internal, pods "inflate-7bfb54665-wzwh2" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.791Z	ERROR	controller.provisioning	Failed to bind default/inflate-85675546d-fct58 to ip-192-168-47-85.ec2.internal, pods "inflate-85675546d-fct58" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.799Z	ERROR	controller.provisioning	Failed to bind default/inflate-85675546d-s5sfb to ip-192-168-47-85.ec2.internal, pods "inflate-85675546d-s5sfb" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.799Z	INFO	controller.provisioning	Bound 0 pod(s) to node ip-192-168-47-85.ec2.internal	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:08.799Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:12.153Z	INFO	controller.provisioning	Batched 30 pods in 3.354438419s	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:12.170Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 13 pod(s) with instance type option(s) [c4.4xlarge c3.4xlarge c6i.4xlarge c5a.4xlarge c5.4xlarge c5ad.4xlarge c5d.4xlarge c5n.4xlarge m5d.4xlarge m5n.4xlarge m6a.4xlarge m5dn.4xlarge m5.4xlarge m5ad.4xlarge m5a.4xlarge m6i.4xlarge m4.4xlarge r3.4xlarge i3.4xlarge r4.4xlarge]	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:14.675Z	INFO	controller.provisioning	Launched instance: i-00e9409082e7e752b, hostname: ip-192-168-68-59.ec2.internal, type: c5n.4xlarge, zone: us-east-1f, capacityType: spot	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:14.707Z	INFO	controller.provisioning	Bound 13 pod(s) to node ip-192-168-68-59.ec2.internal	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:14.707Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:19.273Z	INFO	controller.provisioning	Batched 15 pods in 3.330132281s	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:19.287Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 8 pod(s) with instance type option(s) [m5zn.3xlarge m5dn.4xlarge m5d.4xlarge m5ad.4xlarge m4.4xlarge m6a.4xlarge m5n.4xlarge m5.4xlarge m6i.4xlarge m5a.4xlarge r3.4xlarge]	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.710Z	INFO	controller.provisioning	Launched instance: i-006a9eb958d7a8a64, hostname: ip-192-168-61-229.ec2.internal, type: m6a.4xlarge, zone: us-east-1b, capacityType: spot	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.722Z	ERROR	controller.provisioning	Failed to bind default/inflate-64f6c579c7-wntm9 to ip-192-168-61-229.ec2.internal, pods "inflate-64f6c579c7-wntm9" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.722Z	ERROR	controller.provisioning	Failed to bind default/inflate-64f6c579c7-sl4k2 to ip-192-168-61-229.ec2.internal, pods "inflate-64f6c579c7-sl4k2" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.723Z	ERROR	controller.provisioning	Failed to bind default/inflate-64f6c579c7-4d4wb to ip-192-168-61-229.ec2.internal, pods "inflate-64f6c579c7-4d4wb" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.723Z	ERROR	controller.provisioning	Failed to bind default/inflate-74df46dcf8-kn9jh to ip-192-168-61-229.ec2.internal, pods "inflate-74df46dcf8-kn9jh" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.725Z	ERROR	controller.provisioning	Failed to bind default/inflate-64f6c579c7-jprhs to ip-192-168-61-229.ec2.internal, pods "inflate-64f6c579c7-jprhs" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.726Z	ERROR	controller.provisioning	Failed to bind default/inflate-74df46dcf8-mznj6 to ip-192-168-61-229.ec2.internal, pods "inflate-74df46dcf8-mznj6" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.731Z	ERROR	controller.provisioning	Failed to bind default/inflate-64f6c579c7-rvg2g to ip-192-168-61-229.ec2.internal, pods "inflate-64f6c579c7-rvg2g" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.731Z	ERROR	controller.provisioning	Failed to bind default/inflate-74df46dcf8-twmrs to ip-192-168-61-229.ec2.internal, pods "inflate-74df46dcf8-twmrs" not found	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.731Z	INFO	controller.provisioning	Bound 0 pod(s) to node ip-192-168-61-229.ec2.internal	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:21.731Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:24.774Z	INFO	controller.provisioning	Batched 25 pods in 3.043153549s	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:24.790Z	INFO	controller.provisioning	Computed packing of 1 node(s) for 13 pod(s) with instance type option(s) [c4.4xlarge c3.4xlarge c6i.4xlarge c5.4xlarge c5d.4xlarge c5ad.4xlarge c5a.4xlarge c5n.4xlarge m5.4xlarge m5d.4xlarge m5n.4xlarge m5a.4xlarge m5dn.4xlarge m6a.4xlarge m5ad.4xlarge m6i.4xlarge m4.4xlarge r4.4xlarge i3.4xlarge r3.4xlarge]	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:27.255Z	INFO	controller.provisioning	Launched instance: i-06542cb5ce6f3faa4, hostname: ip-192-168-88-164.ec2.internal, type: c5n.4xlarge, zone: us-east-1f, capacityType: spot	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:27.284Z	INFO	controller.provisioning	Bound 13 pod(s) to node ip-192-168-88-164.ec2.internal	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:46:27.284Z	INFO	controller.provisioning	Waiting for unschedulable pods	{"commit": "df57892", "provisioner": "default"}
2022-02-11T02:47:14.523Z	INFO	controller.node	Added TTL to empty node	{"commit": "df57892", "node": "ip-192-168-47-85.ec2.internal"}
2022-02-11T02:47:25.259Z	INFO	controller.node	Added TTL to empty node	{"commit": "df57892", "node": "ip-192-168-68-59.ec2.internal"}
2022-02-11T02:47:28.225Z	INFO	controller.node	Added TTL to empty node	{"commit": "df57892", "node": "ip-192-168-61-229.ec2.internal"}
2022-02-11T02:47:44.001Z	INFO	controller.node	Triggering termination after 30s for empty node	{"commit": "df57892", "node": "ip-192-168-47-85.ec2.internal"}
2022-02-11T02:47:44.017Z	INFO	controller.termination	Cordoned node	{"commit": "df57892", "node": "ip-192-168-47-85.ec2.internal"}
2022-02-11T02:47:44.242Z	INFO	controller.termination	Deleted node	{"commit": "df57892", "node": "ip-192-168-47-85.ec2.internal"}
2022-02-11T02:47:48.321Z	INFO	controller.node	Removed emptiness TTL from node	{"commit": "df57892", "node": "ip-192-168-61-229.ec2.internal"}
2022-02-11T02:47:48.744Z	INFO	controller.node	Removed emptiness TTL from node	{"commit": "df57892", "node": "ip-192-168-68-59.ec2.internal"}
```

```bash
#!/bin/sh

kubectl delete deployment inflate

launch () {
cat << EOF | kubectl apply -f -
apiVersion: apps/v1
kind: Deployment
metadata:
  name: inflate
spec:
  replicas: 0
  selector:
    matchLabels:
      app: inflate
  template:
    metadata:
      labels:
        app: inflate
    spec:
      terminationGracePeriodSeconds: 0
      containers:
        - name: inflate
          image: public.ecr.aws/eks-distro/kubernetes/pause:3.2
          resources:
            requests:
              cpu: 1
EOF

kubectl scale deployment inflate --replicas=10
}

launch &
for i in `seq 1 30`; do 
  kubectl patch deployment inflate --patch '{"spec": {"template": {"metadata": {"labels": {"test-label": "'$i'"}}}}}'
done
```